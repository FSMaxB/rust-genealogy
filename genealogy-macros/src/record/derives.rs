use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{parse_quote, Path};

pub struct Derives {
	pub equals: bool,
	pub hash: bool,
}

impl Derives {
	pub fn trait_paths(&self) -> Vec<Path> {
		let mut paths: Vec<Path> = vec![parse_quote!(::std::clone::Clone), parse_quote!(::std::fmt::Debug)];
		if self.equals {
			paths.push(parse_quote!(::std::cmp::PartialEq));
			paths.push(parse_quote!(::std::cmp::Eq));
		}

		if self.hash {
			paths.push(parse_quote!(::std::hash::Hash))
		}

		paths
	}
}

impl Default for Derives {
	fn default() -> Self {
		Self {
			equals: true,
			hash: true,
		}
	}
}

impl ToTokens for Derives {
	fn to_tokens(&self, tokens: &mut TokenStream) {
		for trait_path in self.trait_paths() {
			tokens.extend(quote! {
				#[derive(#trait_path)]
			})
		}
	}
}
