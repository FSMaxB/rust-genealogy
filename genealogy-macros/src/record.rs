use crate::attributes::{AttributeProperties, Attributes};
use crate::record::derives::Derives;
use crate::record::fields::RecordFields;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::spanned::Spanned;
use syn::{parse_quote, ItemStruct, Lit};
use syn::{Attribute, Ident, Visibility};

mod derives;
mod field;
mod fields;

pub fn record(record_parameters: TokenStream, item: TokenStream) -> syn::Result<TokenStream> {
	let mut build_constructor = true;
	let mut derives = Derives::default();
	if !record_parameters.is_empty() {
		let record_attributes = syn::parse2::<AttributeProperties>(record_parameters)?;

		for (property, value) in record_attributes.properties {
			let name = property.to_string();
			match (name.as_str(), value) {
				("constructor", Lit::Bool(boolean)) => {
					build_constructor = boolean.value;
				}
				("equals", Lit::Bool(boolean)) => {
					derives.equals = boolean.value;
				}
				("hash", Lit::Bool(boolean)) => {
					derives.hash = boolean.value;
				}
				_ => {
					return Err(syn::Error::new(
						property.span(),
						r#"Invalid attribute property for #[record], allowed are "constructor", "equals" and "hash" with a boolean value."#,
					));
				}
			}
		}
	}

	let ItemStruct {
		attrs,
		vis,
		struct_token: _,
		ident,
		generics,
		fields,
		semi_token: _,
	} = syn::parse2::<ItemStruct>(item)?;

	if generics.gt_token.is_some() || generics.lt_token.is_some() || generics.where_clause.is_some() {
		return Err(syn::Error::new(
			generics.span(),
			"#[record] currently doesn't work with generics.",
		));
	}

	let mut required_traits = derives.trait_paths();
	required_traits.push(parse_quote!(::std::fmt::Display));

	let fields = RecordFields::new(fields, vis.clone())?;
	let mut tokens = TokenStream::new();
	fields.to_type_assertions(&required_traits, &mut tokens);

	let record = Record {
		attributes: attrs.into(),
		visibility: vis,
		name: ident,
		fields,
		build_constructor,
		derives,
	};

	record.to_tokens(&mut tokens);

	Ok(tokens)
}

struct Record {
	attributes: Attributes,
	visibility: Visibility,
	name: Ident,
	fields: RecordFields,
	build_constructor: bool,
	derives: Derives,
}

impl ToTokens for Record {
	fn to_tokens(&self, tokens: &mut TokenStream) {
		let Self {
			attributes,
			visibility,
			name,
			fields,
			build_constructor,
			derives,
		} = self;

		let mut items = TokenStream::new();
		fields.to_items(&mut items);

		let constructor = if *build_constructor {
			let mut parameters = TokenStream::new();
			fields.to_parameters(&mut parameters);
			let mut initializers = TokenStream::new();
			fields.to_initializers(&mut initializers);
			quote! {
				#visibility fn new(#parameters) -> Self {
					Self {
						#initializers
					}
				}
			}
		} else {
			quote!()
		};

		let mut accessors = TokenStream::new();
		fields.to_accessors(&mut accessors);

		let mut display = TokenStream::new();
		fields.to_display_implementation(&self.name, &mut display);

		tokens.extend(quote! {
			#derives
			#attributes
			#visibility struct #name {
				#items
			}

			impl #name {
				#constructor

				#accessors
			}

			#display
		});
	}
}

fn is_omit(attribute: &Attribute) -> bool {
	let path = attribute
		.path
		.segments
		.iter()
		.map(|segment| &segment.ident)
		.collect::<Vec<_>>();
	match path.as_slice() {
		[identifier] => *identifier == "omit",
		_ => false,
	}
}
