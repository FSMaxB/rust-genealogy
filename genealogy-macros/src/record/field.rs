use crate::attributes::Attributes;
use crate::record::is_omit;
use proc_macro2::TokenStream;
use quote::quote;
use syn::spanned::Spanned;
use syn::{Field, Ident, Type, Visibility};

pub struct RecordField {
	pub attributes: Attributes,
	pub visibility: Visibility,
	pub name: Ident,
	pub r#type: Type,
	pub omit: bool,
}

impl RecordField {
	pub fn new(field: Field, visibility: Visibility) -> syn::Result<RecordField> {
		let field_span = field.span();
		let mut omit = false;
		let attributes = field
			.attrs
			.into_iter()
			.filter(|attribute| {
				if is_omit(attribute) {
					omit = true;
					false
				} else {
					true
				}
			})
			.collect::<Vec<_>>()
			.into();

		let name = match field.ident {
			Some(name) => name,
			None => {
				return Err(syn::Error::new(field_span, "#[record] doesn't support tuple structs"));
			}
		};
		match field.vis {
			Visibility::Inherited => {}
			visibility => {
				return Err(syn::Error::new(
					visibility.span(),
					"All fields in #[record] must be private",
				));
			}
		}

		Ok(Self {
			attributes,
			visibility,
			name,
			r#type: field.ty,
			omit,
		})
	}

	pub fn to_item(&self, tokens: &mut TokenStream) {
		let Self {
			attributes,
			name,
			r#type,
			..
		} = self;
		tokens.extend(quote! {
			#attributes
			#name: #r#type,
		})
	}

	pub fn to_parameter(&self, tokens: &mut TokenStream) {
		let Self { name, r#type, .. } = self;
		tokens.extend(quote! {
			#name: #r#type,
		});
	}

	pub fn to_initializer(&self, tokens: &mut TokenStream) {
		let Self { name, .. } = self;
		tokens.extend(quote!(#name,));
	}

	pub fn to_accessor(&self, tokens: &mut TokenStream) {
		let Self {
			visibility,
			name,
			r#type,
			omit,
			..
		} = self;

		if *omit {
			return;
		}

		tokens.extend(quote! {
			#visibility fn #name(&self) -> #r#type {
				::std::clone::Clone::clone(&self.#name)
			}
		});
	}
}
