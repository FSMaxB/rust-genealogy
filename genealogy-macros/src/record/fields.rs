use crate::record::field::RecordField;
use proc_macro2::TokenStream;
use quote::quote;
use syn::punctuated::Punctuated;
use syn::spanned::Spanned;
use syn::{parse_quote, ExprField, Fields, FieldsNamed, Ident, Path, Token, Visibility};

pub struct RecordFields(Vec<RecordField>);

impl RecordFields {
	pub fn new(fields: Fields, record_visibility: Visibility) -> syn::Result<RecordFields> {
		let fields = match fields {
			Fields::Named(FieldsNamed { named, .. }) => named,
			_ => {
				return Err(syn::Error::new(
					fields.span(),
					"#[record] is only allowed for structs with named fields.",
				))
			}
		};
		let record_fields = fields
			.into_iter()
			.map(move |field| RecordField::new(field, record_visibility.clone()))
			.collect::<Result<_, _>>()?;
		Ok(RecordFields(record_fields))
	}

	pub fn to_items(&self, tokens: &mut TokenStream) {
		for field in &self.0 {
			field.to_item(tokens);
		}
	}

	pub fn to_parameters(&self, tokens: &mut TokenStream) {
		for field in &self.0 {
			field.to_parameter(tokens);
		}
	}

	pub fn to_initializers(&self, tokens: &mut TokenStream) {
		for field in &self.0 {
			field.to_initializer(tokens);
		}
	}

	pub fn to_accessors(&self, tokens: &mut TokenStream) {
		for field in &self.0 {
			field.to_accessor(tokens);
		}
	}

	pub fn to_display_implementation(&self, record_name: &Ident, tokens: &mut TokenStream) {
		let field_format = self
			.0
			.iter()
			.map(|field| format!("{}={{}}", field.name))
			.collect::<Vec<_>>();
		let format_string = format!("{}[{}]", record_name, field_format.join(", "));

		let format_arguments = self
			.0
			.iter()
			.map(|field| -> ExprField {
				let name = &field.name;
				parse_quote!(self.#name)
			})
			.collect::<Punctuated<_, Token!(,)>>();

		tokens.extend(quote! {
			impl ::std::fmt::Display for #record_name {
				fn fmt(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
					::std::write!(formatter, #format_string, #format_arguments)
				}
			}
		})
	}

	pub fn to_type_assertions(&self, trait_paths: &[Path], tokens: &mut TokenStream) {
		for field in &self.0 {
			field.to_type_assertion(trait_paths, tokens);
		}
	}
}
