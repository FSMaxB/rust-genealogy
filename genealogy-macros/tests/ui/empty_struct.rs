use genealogy_macros::record;

#[record]
struct Record;

pub fn main() {}
