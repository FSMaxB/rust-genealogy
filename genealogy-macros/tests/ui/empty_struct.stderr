error: #[record] is only allowed for structs with named fields.
 --> tests/ui/empty_struct.rs:3:1
  |
3 | #[record]
  | ^^^^^^^^^
  |
  = note: this error originates in the attribute macro `record` (in Nightly builds, run with -Z macro-backtrace for more info)
